File: language/transaction-builder/generator/examples/cpp/stdlib_demo.cpp
// Copyright (c) The Diem Core Contributors
// SPDX-License-Identifier: Apache-2.0

#include "diem_framework.hpp"
#include "diem_types.hpp"
#include <memory>

using namespace diem_framework;
using namespace diem_types;

void demo_p2p_script() {
    auto token = TypeTag{TypeTag::Struct{StructTag{
        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1},
        {"XDX"},
        {"XDX"},
        {},
    }}};
    auto payee = AccountAddress{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22,
                                0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22};
    uint64_t amount = 1234567;
    auto script =
        encode_peer_to_peer_with_metadata_script(token, payee, amount, {}, {});

    auto output = script.bcsSerialize();
    for (uint8_t o : output) {
        printf("%d ", o);
    };
    printf("\n");
}

void demo_p2p_script_function() {
    auto token = TypeTag{TypeTag::Struct{StructTag{
        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1},
        {"XDX"},
        {"XDX"},
        {},
    }}};
    auto payee = AccountAddress{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22,
                                0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22};
    uint64_t amount = 1234567;
    auto payload =
        encode_peer_to_peer_with_metadata_script_function(token, payee, amount, {}, {});

    auto output = payload.bcsSerialize();
    for (uint8_t o : output) {
        printf("%d ", o);
    };
    printf("\n");
}

int main() {
    demo_p2p_script();
    demo_p2p_script_function();
    return 0;
}

File: language/transaction-builder/generator/examples/python3/stdlib_demo.py
# Copyright (c) The Diem Core Contributors
# SPDX-License-Identifier: Apache-2.0

# pyre-strict

import diem_types as diem
import serde_types as st
import diem_framework as stdlib


def demo_p2p_script() -> None:
    token = diem.TypeTag__Struct(
        value=diem.StructTag(
            address=diem.AccountAddress.from_bytes(b"\x00" * 15 + b"\x01"),
            module=diem.Identifier("XDX"),
            name=diem.Identifier("XDX"),
            type_params=[],
        )
    )
    payee = diem.AccountAddress.from_bytes(b"\x22" * 16)
    amount = st.uint64(1_234_567)
    script = stdlib.encode_peer_to_peer_with_metadata_script(token, payee, amount, b"", b"")

    call = stdlib.decode_script(script)
    assert isinstance(call, stdlib.ScriptCall__PeerToPeerWithMetadata)
    assert call.amount == amount;
    assert call.payee == payee;

    for b in script.bcs_serialize():
        print("%d " % b, end='')
    print()

def demo_p2p_script_function() -> None:
    token = diem.TypeTag__Struct(
        value=diem.StructTag(
            address=diem.AccountAddress.from_bytes(b"\x00" * 15 + b"\x01"),
            module=diem.Identifier("XDX"),
            name=diem.Identifier("XDX"),
            type_params=[],
        )
    )
    payee = diem.AccountAddress.from_bytes(b"\x22" * 16)
    amount = st.uint64(1_234_567)
    payload = stdlib.encode_peer_to_peer_with_metadata_script_function(token, payee, amount, b"", b"")

    call = stdlib.decode_script_function_payload(payload)
    assert isinstance(call, stdlib.ScriptFunctionCall__PeerToPeerWithMetadata)
    assert call.amount == amount;
    assert call.payee == payee;

    for b in payload.bcs_serialize():
        print("%d " % b, end='')
    print()

if __name__ == "__main__":
    demo_p2p_script()
    demo_p2p_script_function()
